File: 1358-number-of-substrings-containing-all-three-characters/1358-number-of-substrings-containing-all-three-characters.cs
public class Solution {
    public int NumberOfSubstrings(string s) {
        int n = s.Length, count = 0;
        var freq = new int[3];

        for(int l = 0, r = 0; r < n; r++){
            freq[s[r] - 'a']++;
            // every time we find all a,b,c - we can expand it to end of string potentially
            while(freq[0] > 0 && freq[1] > 0 && freq[2] > 0){
                count += n - r;
                freq[s[l] - 'a']--;
                l++;
            }
        }

        return count;
    }
}

File: 1358-number-of-substrings-containing-all-three-characters/1358-number-of-substrings-containing-all-three-characters.py
class Solution:
    def numberOfSubstrings(self, s: str) -> int:
        n, count = len(s), 0
        freq = [0] * 3

        a = ord('a')
        l = 0
        for r in range(0, n):
            freq[ord(s[r]) - a] += 1
            # every time we find all a,b,c - we can expand it to end of string potentially
            while freq[0] > 0 and freq[1] > 0 and freq[2] > 0:
                count += n - r
                freq[ord(s[l]) - a] -= 1
                l += 1

        return count